Components/LearningCore/Source/gov/sandia/cognition/learning/algorithm/clustering/OptimizedAffinityPropagation.java

```diff
@@ -0,0 +1,318 @@
+      this.setClusters(new HashMap<>());
+      for (int i = 0; i < exampleCount; i++)
+      {
+        this.assignments[i] = -1;
+      }
+
+      // Ready to learn.
+      return true;
+    }
+
+    /**
+     * Set the self similarity based on the self divergence.
+     */
+    private void fillSimilarityDiagonal()
+    {
+      for (int i = 0; i < exampleCount; i++)
+      {
+        similarities[i][i] = -getSelfDivergence();
+      }
+    }
+
+    private void fillSimilarityStructures(double[] divergenciesArray)
+    {
+      for (int i = 0; i < exampleCount; i++)
+      {
+        final T exampleI = examples.get(i);
+
+        for (int j = 0; j <= i; j++)
+        {
+          // We compute similarity, which is the negative of divergence,
+          // since a lower divergence means a higher similarity.
+          final T exampleJ = examples.get(j);
+          final double similarity = -divergence.evaluate(
+              exampleI, exampleJ);
+          similarities[i][j] = similarity;
+          similarities[j][i] = similarity;
+          divergenciesArray[i * (i+1) / 2 + j] = -similarity;
+        }
+      }
+    }
+
+    private double computeMedian(double[] divergenciesArray)
+    {
+
+      double[] doubles = Arrays.stream(divergenciesArray)
+          .flatMap(DoubleStream::of)
+          .distinct()
+          .sorted()
+          .toArray();
+
+      double median;
+      int size = doubles.length;
+      if (size == 0) {
+        median = 0;
+      } else {
+        if (size % 2 == 0)
+          median = (doubles[size / 2] + doubles[size / 2 - 1]) / 2;
+        else
+          median = doubles[size / 2];
+      }
+      return median;
+
+    }
+
+    /**
+     * Updates the responsibilities matrix using the similarity values and the
+     * current availability values.
+     */
+    @Override
+    protected void updateResponsibilities()
+    {
+
+      for(int i = 0 ; i < exampleCount; i++) {
+        double[] rowArray = createSumArray(availabilities[i], similarities[i]);
+        MaxResult maxResult = computeMax(rowArray);
+        for (int j = 0; j < exampleCount; j++) {
+          double maxToSubtract = j != maxResult.maxIndex ? maxResult.maxValue : maxResult.secondMaxValue;
+          responsibilities[i][j] = oneMinusDampingFactor * (similarities[i][j] - maxToSubtract) + dampingFactor * responsibilities[i][j];
+        }
+      }
+    }
+
+    private MaxResult computeMax(double[] rowArray)
+    {
+      MaxResult maxResult = new MaxResult();
+      for (int i = 0; i < rowArray.length; i++) {
+        if(rowArray[i] > maxResult.maxValue){
+          maxResult.secondMaxValue = maxResult.maxValue;
+          maxResult.maxValue = rowArray[i];
+          maxResult.maxIndex = i;
+        }else if(rowArray[i] > maxResult.secondMaxValue){
+          maxResult.secondMaxValue = rowArray[i];
+        }
+      }
+      return maxResult;
+    }
+
+
+    private double[] createSumArray(double[] availability, double[] similarity)

```

Comment:
You may be able to avoid needing a utility function like this if you use DenseVectors instead of doing it all with raw arrays. There may be a slight performance penalty from the vector overhead, though that is generally what we use in the rest of the foundry to get the easier utility methods.

Reply:
I don't get if you would like to change the raw arrays structure wherever or only in this point. I'm following what is already implemented in AffinityPropagation class. If i have to change it, could you provide me an example, please?

Reply:
Its true those classes use the raw arrays, but they don't need to do these kinds of vector-style operations in their implementation. The way to do it would be to use the DenseVector via VectorFactory.getDenseDefault to make the two arrays, then fill them via set(), then this operation is just availability.plus(similarity).
But that may require changing quite a bit of the code, so if you think it is better to keep it closer to the other implementation then you can leave it this way (though may want to add a note saying why).